pkg/test/framework/components/cluster/aggregate/factory.go

```diff
@@ -0,0 +1,148 @@
+
+func NewFactory() Factory {
+	return &aggregateFactory{
+		factories:   make(map[cluster.Kind]cluster.Factory),
+		allClusters: map[string]resource.Cluster{},
+	}
+}
+
+type aggregateFactory struct {
+	allClusters map[string]resource.Cluster
+	configs     []cluster.Config
+	factories   map[cluster.Kind]cluster.Factory
+}
+
+func (a *aggregateFactory) Kind() cluster.Kind {
+	return cluster.Aggregate
+}
+
+func (a *aggregateFactory) With(configs ...cluster.Config) cluster.Factory {
+	a.configs = append(a.configs, configs...)
+	for _, config := range a.configs {
+		a.maybeCreateFactory(config)
+	}
+	return a
+}
+
+func (a *aggregateFactory) Build() (resource.Clusters, error) {
+	scopes.Framework.Infof("=== BEGIN: Building clusters ===")
+
+	var errs error
+	// distribute configs to their factories
+	for i, origCfg := range a.configs {
+		cfg, err := validConfig(origCfg)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+		factory, ok := a.factories[origCfg.Kind]
+		if !ok {
+			errs = multierror.Append(errs, fmt.Errorf("invalid kind for %s (item %d): %s", cfg.Name, i, cfg.Kind))
+			continue
+		}
+		factory.With(cfg)
+	}
+
+	// initialize the clusters
+	var clusters resource.Clusters
+	for kind, factory := range a.factories {
+		scopes.Framework.Infof("Building %s clusters", kind)
+		built, err := factory.Build()
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+		for _, c := range built {
+			if _, ok := a.allClusters[c.Name()]; ok {
+				errs = multierror.Append(errs, fmt.Errorf("duplicate cluster name: %s", c.Name()))
+				continue
+			}
+			a.allClusters[c.Name()] = c
+			scopes.Framework.Infof(c.String())
+		}
+		clusters = append(clusters, built...)
+	}
+	if errs != nil {
+		scopes.Framework.Infof("=== FAILED: Building clusters ===")
+		return nil, errs
+	}
+	for n, c := range a.allClusters {
+		scopes.Framework.Infof("Built Cluster: %s", n)
+		scopes.Framework.Infof(c.String())
+	}
+	scopes.Framework.Infof("=== DONE: Building clusters ===")
+
+	return clusters, errs
+}
+
+func (a *aggregateFactory) BuildOrFail(t test.Failer) resource.Clusters {
+	out, err := a.Build()
+	if err != nil {
+		t.Fatal(err)
+	}
+	return out
+}
+
+func validConfig(cfg cluster.Config) (cluster.Config, error) {
+	if cfg.Name == "" {
+		return cfg, fmt.Errorf("empty cluster name")
+	}
+	if cfg.Kind == "" {
+		return cfg, fmt.Errorf("unspecified Kind for %s", cfg.Name)
+	}
+	if cfg.ControlPlaneClusterName == "" {
+		cfg.ControlPlaneClusterName = cfg.Name
+	}
+	if cfg.ConfigClusterName == "" {
+		cfg.ConfigClusterName = cfg.Name
+	}
+	return cfg, nil
+}
+
+// maybeCreateFactory initializes concrete factory implementations.
+// If the given Kind is unsupported we err during the build step to allow collecting
+// as much validation info as possible.
+func (a *aggregateFactory) maybeCreateFactory(config cluster.Config) {
+	switch config.Kind {
+	case cluster.Kubernetes:
+		if a.factories[cluster.Kubernetes] == nil {
```

Comment:
Rather than hard-coding the list of factories here, might it be better to have the factories just register with this central registry? So in the `kube` package, for example, you'd have something like:

```golang
func init() {
  registry.Add(&factory)
}
```

I think this would simplify the process for expanding the factory list. WDYT?

Reply:
Sounds good. Maybe it should register a parameter-less constructor to add some extra immutability guarantees.

Reply:
Nvm, let's just make immutability a hard requirement.